src/nvim/msgpack_rpc/channel.c

```diff
@@ -321,26 +321,31 @@ static void parse_msgpack(Channel *channel)
       }
       arena_mem_free(arena_finish(&p->arena));
     } else if (p->type == kMessageTypeResponse) {
-      ChannelCallFrame *frame = kv_last(channel->rpc.call_stack);
-      if (p->request_id != frame->request_id) {
-        char buf[256];
-        snprintf(buf, sizeof(buf),
-                 "ch %" PRIu64 " returned a response with an unknown request "
-                 "id. Ensure the client is properly synchronized",
-                 channel->id);
-        chan_close_with_error(channel, buf, LOGLVL_ERR);
-      }
-      frame->returned = true;
-      frame->errored = (p->error.type != kObjectTypeNil);
-
-      if (frame->errored) {
-        frame->result = p->error;
-        // TODO(bfredl): p->result should not even be decoded
-        // api_free_object(p->result);
-      } else {
-        frame->result = p->result;
+      // Find the corresponding frame of the response
+      for (size_t i = 0; i < kv_size(channel->rpc.call_stack); i++) {
+        ChannelCallFrame *frame = kv_Z(channel->rpc.call_stack, i);
+        if (p->request_id != frame->request_id) {
+          continue;
+        }
+        frame->returned = true;
+        frame->errored = (p->error.type != kObjectTypeNil);
+        if (frame->errored) {
+          frame->result = p->error;
+          // TODO(bfredl): p->result should not even be decoded
+          // api_free_object(p->result);
+        } else {
+          frame->result = p->result;
+        }
+        frame->result_mem = arena_finish(&p->arena);
+        goto end;
       }
-      frame->result_mem = arena_finish(&p->arena);
+      // No corresponding frame found for the response
+      char buf[256];
+      snprintf(buf, sizeof(buf),
+                "ch %" PRIu64 " returned a response with an unknown request "
+                "id %" PRIu32 ". Ensure the client is properly synchronized",
+                channel->id, p->request_id);
+      chan_close_with_error(channel, buf, LOGLVL_ERR);
```

Comment:
I think we should just ignore this case while it is likely caused by duplicated messages.

Reply:
Duplicated messages, out-of-order messages... aren't those an indication that something is wrong? If so, silently ignoring them makes things worse. 

Why are messages being duplicated? Should we not at least _log_ such a case?

Reply:
> Why are messages being duplicated? Should we not at least log such a case (perhaps at WARN level)? Silently ignoring problems leads to very difficult bugs.

Totally agree. Yes, I said "just ignore" but I didn't mean just ignore, of course 🙇